File: 2026/07-July/29-Wednesday/solution.java
import java.util.*;

class Solution {
    private static final int MAX = 1_000_001;

    public String smallestPalindrome(String s, int k) {
        int[] freq = new int[26];

        for (char c : s.toCharArray()) {
            freq[c - 'a']++;
        }

        int[] half = new int[26];
        String mid = "";

        for (int i = 0; i < 26; i++) {
            half[i] = freq[i] / 2;
            if ((freq[i] & 1) == 1) {
                mid = String.valueOf((char) ('a' + i));
            }
        }

        if (countWays(half) < k) {
            return "";
        }

        StringBuilder left = new StringBuilder();
        int len = 0;
        for (int x : half) len += x;

        for (int pos = 0; pos < len; pos++) {
            for (int i = 0; i < 26; i++) {
                if (half[i] == 0) continue;

                half[i]--;
                int ways = countWays(half);

                if (ways >= k) {
                    left.append((char) ('a' + i));
                    break;
                }

                k -= ways;
                half[i]++;
            }
        }

        StringBuilder ans = new StringBuilder();
        ans.append(left);
        ans.append(mid);
        ans.append(left.reverse());

        return ans.toString();
    }

    private int countWays(int[] cnt) {
        int total = 0;
        for (int x : cnt) total += x;

        long res = 1;

        for (int f : cnt) {
            res *= nCk(total, f);
            if (res >= MAX) return MAX;
            total -= f;
        }

        return (int) res;
    }

    private long nCk(int n, int k) {
        k = Math.min(k, n - k);
        long res = 1;

        for (int i = 1; i <= k; i++) {
            res = res * (n - i + 1) / i;
            if (res >= MAX) return MAX;
        }

        return res;
    }
}

File: 2026/07-July/29-Wednesday/solution.py
from collections import Counter

class Solution:
    def __init__(self):
        self.MAX = 10**6 + 1

    def smallestPalindrome(self, s: str, k: int) -> str:
        count = Counter(s)

        halfCount = [0] * 26
        midLetter = ""

        for c, freq in count.items():
            halfCount[ord(c) - ord("a")] = freq // 2
            if freq % 2:
                midLetter = c

        if self._countArrangements(halfCount) < k:
            return ""

        left = []
        halfLen = sum(halfCount)

        for _ in range(halfLen):
            for i in range(26):
                if halfCount[i] == 0:
                    continue

                halfCount[i] -= 1
                ways = self._countArrangements(halfCount)

                if ways >= k:
                    left.append(chr(i + ord("a")))
                    break

                k -= ways
                halfCount[i] += 1

        return "".join(left) + midLetter + "".join(reversed(left))

    def _countArrangements(self, cnt):
        total = sum(cnt)
        res = 1

        for f in cnt:
            res *= self._nCk(total, f)
            if res >= self.MAX:
                return self.MAX
            total -= f

        return res

    def _nCk(self, n, k):
        k = min(k, n - k)
        res = 1

        for i in range(1, k + 1):
            res = res * (n - i + 1) // i
            if res >= self.MAX:
                return self.MAX

        return res
